File: app/src/main/java/com/gmail/cristiandeives/motofretado/http/jsonapi/BusData.java
package com.gmail.cristiandeives.motofretado.http.jsonapi;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.gmail.cristiandeives.motofretado.http.Bus;

class BusData {
    @Nullable String type;
    @Nullable String id;
    @Nullable BusAttributes attributes;

    static @NonNull BusData parse(@NonNull Bus bus) {
        BusData data = new BusData();

        data.id = bus.id;
        data.type = BusDocument.TYPE;
        data.attributes = BusAttributes.parse(bus);

        return data;
    }

    @NonNull Bus toBus() throws IllegalStateException {
        if (type == null) {
            throw new IllegalStateException("missing JSONAPI type");
        }

        if (!type.equals(BusDocument.TYPE)) {
            throw new IllegalStateException("invalid JSONAPI type: " + type);
        }

        if (TextUtils.isEmpty(id)) {
            throw new IllegalStateException("missing JSONAPI ID");
        }

        Bus bus = new Bus();

        bus.id = id;
        if (attributes != null) {
            bus.createdAt = attributes.createdAt;
            bus.updatedAt = attributes.updatedAt;
            bus.latitude = attributes.latitude;
            bus.longitude = attributes.longitude;
        }

        return bus;
    }
}

File: app/src/main/java/com/gmail/cristiandeives/motofretado/http/jsonapi/BusData.kt
package com.gmail.cristiandeives.motofretado.http.jsonapi

import com.gmail.cristiandeives.motofretado.http.Bus

internal data class BusData(val type: String, val id: String, val attributes: BusAttributes?) {
    internal constructor(bus: Bus) : this(BusDocument.TYPE, bus.id, BusAttributes(bus))

    internal fun toBus(): Bus {
        if (type != BusDocument.TYPE) {
            throw IllegalStateException("invalid JSONAPI type: $type")
        }

        if (id.isEmpty()) {
            throw IllegalStateException("missing JSONAPI ID")
        }

        return Bus(
                id,
                attributes?.latitude,
                attributes?.longitude,
                attributes?.createdAt,
                attributes?.updatedAt
        )
    }
}